### PR TITLE
Centralizing Partitioning State

### PR DESCRIPTION
**Pull request by narendasan**
# Description

It is incredibly hard to track the state of the partitioning process because the pipeline relies on a number of shared lookup tables that are modified in place during the course of partitioning. 

`PartitioningCtx`, similar to `ConversionCtx` centralizes key information like decisions about node executors, user settings, the in progress graph etc. so that this information is uniformly managed and queryable at any point during the pipelines execution. This means it is hard to client code to alter state in unpredictable ways and it also reduces the amount of arguments that need to be passed to each phase. 

Addresses #949 

## Type of change

Please delete options that are not relevant and/or add your own.

- New feature (non-breaking change which adds functionality)

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

**Comment by narendasan** [comment]
@bowang007 This should be roughly equivalent to our current system. There are a couple improvements on this that I'd like to do that I want your input on. 

1. I renamed a bunch of stuff to hopefully help make it clearer to new devs, let me know if you disagree about anything. 
2. I want every node in the `global_fallback_map` (now called `node_executor_decision_map`) and make an explicit decision on if it will run in torch or tensorrt (this can be changed like if a block would run in trt but is not large enough). That way in debugging or otherwise I can just ask about a particular node and get back what it will run on. Right now I am having trouble doing this because there is an assumption in the system that if the node is not in the map it must run in tensorrt. 
3. There is still work to be done on being able to easily dump state at any given time. 
4. Are there other data sources that we might want to have managed by the context?

File: core/partitioning/partitioningctx/PartitioningCtx.cpp [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    return false;
+  } else {
+    return true;
+  }
+}
+
+std::ostream& operator<<(std::ostream& os, const NodeExecutorDecision& format) {
+  switch (format) {
+    case NodeExecutorDecision::kUNSUPPORTED:
+      return os << "to run torch due to lack of converter support";
```
Comment (narendasan) [inline core/partitioning/partitioningctx/PartitioningCtx.cpp:95]: "run in torch"

File: core/compiler.cpp [hunk 2]
```diff
@@ -272,8 +269,8 @@ GraphAndMapping ConstructFallbackGraph(
         // convert the 2 blocks in prim::if and get the converted graph with mappings
         std::vector<GraphAndMapping> graph_and_mappings;
         for (auto cur_block : if_node->blocks()) {
-          graph_and_mappings.push_back(
-              ConstructFallbackGraph(new_mod, cur_block, example_tensor_map, cfg, static_params, fallback_nodes));
+          graph_and_mappings.push_back(ConstructFallbackGraph_(
```
Comment (narendasan) [inline core/compiler.cpp:272]: Not wildly pressing but is there a way to do all the partitioning beforehand then go through and compile specific blocks? Having them mixed is not as easy to debug

File: core/compiler.cpp [hunk 3]
```diff
@@ -219,19 +219,16 @@ void AddIfBlockToGraph(
   return;
 }
 
-GraphAndMapping ConstructFallbackGraph(
+GraphAndMapping ConstructFallbackGraph_(
```
Comment (narendasan) [inline core/compiler.cpp:222]: Have a more distinguishing name for this maybe?

File: core/compiler.cpp (hunk 2 above)
Comment (narendasan) [reply core/compiler.cpp:272]: @bowang007 thoughts here?

Comment (narendasan) [reply core/compiler.cpp:272]: Perhaps what we do is recursively partition, then recursively compile the final graph. Not sure if graph stitching can handle this right now.

**Comment by bowang007** [comment]
> @bowang007 This should be roughly equivalent to our current system. There are a couple improvements on this that I'd like to do that I want your input on.
> 
> 1. I renamed a bunch of stuff to hopefully help make it clearer to new devs, let me know if you disagree about anything.
> 2. I want every node in the `global_fallback_map` (now called `node_executor_decision_map`) and make an explicit decision on if it will run in torch or tensorrt (this can be changed like if a block would run in trt but is not large enough). That way in debugging or otherwise I can just ask about a particular node and get back what it will run on. Right now I am having trouble doing this because there is an assumption in the system that if the node is not in the map it must run in tensorrt.
> 3. There is still work to be done on being able to easily dump state at any given time.
> 4. Are there other data sources that we might want to have managed by the context?

Went through just now, I think the general architecture is very clear and comprehensive, here are some thoughts:
2. We can have a map for every nodes in the graph, logic might be a little bit different but shouldn't be to complicated. 
3. more details?

**Comment by narendasan** [comment]
> > @bowang007 This should be roughly equivalent to our current system. There are a couple improvements on this that I'd like to do that I want your input on.
> > 
> > 1. I renamed a bunch of stuff to hopefully help make it clearer to new devs, let me know if you disagree about anything.
> > 2. I want every node in the `global_fallback_map` (now called `node_executor_decision_map`) and make an explicit decision on if it will run in torch or tensorrt (this can be changed like if a block would run in trt but is not large enough). That way in debugging or otherwise I can just ask about a particular node and get back what it will run on. Right now I am having trouble doing this because there is an assumption in the system that if the node is not in the map it must run in tensorrt.
> > 3. There is still work to be done on being able to easily dump state at any given time.
> > 4. Are there other data sources that we might want to have managed by the context?
> 
> Went through just now, I think the general architecture is very clear and comprehensive, here are some thoughts: 2. We can have a map for every nodes in the graph, logic might be a little bit different but shouldn't be to complicated. 3. more details on this?

I want to be able to enable features like https://github.com/pytorch/TensorRT/discussions/1257
which would be able to show a user info like inputs to a block, node assignments, shape calculation results etc.

**Comment by bowang007** [comment]
@narendasan If we store SegmentedBlocks https://github.com/pytorch/TensorRT/blob/1a4084b39a4725cf4818c2ebcab455cd1793031e/core/partitioning/partitioningctx/PartitioningCtx.h#L51  in PartitioningCtx, what about the sub-blocks' segmentedBlocks? Looks like it could get messed up if there is some node like If::node which contains sub-blocks.

**Comment by narendasan** [comment]
> @narendasan If we store SegmentedBlocks
> 
> https://github.com/pytorch/TensorRT/blob/1a4084b39a4725cf4818c2ebcab455cd1793031e/core/partitioning/partitioningctx/PartitioningCtx.h#L51
> in PartitioningCtx, what about the sub-blocks' segmentedBlocks? Looks like it could get messed up if there is some node like If::node which contains sub-blocks.

Where do we put sub blocks now?

**Comment by bowang007** [comment]
> > @narendasan If we store SegmentedBlocks
> > https://github.com/pytorch/TensorRT/blob/1a4084b39a4725cf4818c2ebcab45 5cd1793031e/core/partitioning/partitioningctx/PartitioningCtx.h#L51
> > 
> > in PartitioningCtx, what about the sub-blocks' segmentedBlocks? Looks like it could get messed up if there is some node like If::node which contains sub-blocks.
> 
> Where do we put sub blocks now?

We are doing recursion for sub-blocks, it that stack ends then the sub-block will be stitched and put back to the if nodes.

**Comment by bowang007** [comment]
> @bowang007 This should be roughly equivalent to our current system. There are a couple improvements on this that I'd like to do that I want your input on.
> 
> 1. I renamed a bunch of stuff to hopefully help make it clearer to new devs, let me know if you disagree about anything.
> 2. I want every node in the `global_fallback_map` (now called `node_executor_decision_map`) and make an explicit decision on if it will run in torch or tensorrt (this can be changed like if a block would run in trt but is not large enough). That way in debugging or otherwise I can just ask about a particular node and get back what it will run on. Right now I am having trouble doing this because there is an assumption in the system that if the node is not in the map it must run in tensorrt.
> 3. There is still work to be done on being able to easily dump state at any given time.
> 4. Are there other data sources that we might want to have managed by the context?

For 4: I'm now trying to peel off the stitching part out of compiler.cpp so the LUT for mapping between old value and new value might be stored in the context as well.

**Pull request merged**